finetuner/tuner/callback/best_model_checkpoint.py

```diff
@@ -0,0 +1,140 @@
+from jina.logging.logger import JinaLogger
+
+from finetuner.helper import get_framework
+from .base import BaseCallback
+
+
+if TYPE_CHECKING:
+    from ..base import BaseTuner
+
+
+class BestModelCheckpoint(BaseCallback):
+    """
+    Callback to save model at every epoch or the best model across all epochs
+    `BestModelCheckpoint` is used in conjunction with training
+    using `finetuner.fit()`
+
+    An option this callback provides include:
+    - Definition of 'best'; which quantity to monitor and whether it should be
+        maximized or minimized.
+    """
+
+    def __init__(
+        self,
+        save_dir: str = None,
+        monitor: str = 'val_loss',
+        mode: Optional[str] = None,
+    ):
+        """
+        :param save_dir: string or `PathLike`, path to save the model file.
+        :param monitor: if `monitor='loss'` best bodel saved will be according
+            to the training loss, if `monitor='val_loss'` best model saved will be
+            according to the validation loss
+        :param mode: one of {'auto', 'min', 'max'}. the
+            decision to overwrite the current save file is made based on either
+            the maximization or the minimization of the monitored quantity.
+            For `val_acc`, this should be `max`, for `val_loss` this should be
+            `min`, etc. In `auto` mode, the mode is set to `max` if the quantities
+            monitored are 'acc' or start with 'fmeasure' and are set to `min` for
+            the rest of the quantities.
+        """
+        self._logger = JinaLogger(self.__class__.__name__)
+        self._save_dir = save_dir
+        self._monitor = monitor
+        self._train_losses = []
+        self._valid_losses = []
+        if not save_dir:
+            raise ValueError(
+                '``save_dir`` parameter is mandatory. Pass it in parameters'
+            )
+
+        if mode not in ['auto', 'min', 'max']:
+            self._logger.warning(
+                'ModelCheckpoint mode %s is unknown, ' 'fallback to auto mode.', mode
+            )
+            mode = 'auto'
+
+        if mode == 'min':
+            self._monitor_op = np.less
+            self._best = np.Inf
+        elif mode == 'max':
+            self._monitor_op = np.greater
+            self._best = -np.Inf
+        else:
+            if 'acc' in self._monitor or self._monitor.startswith('fmeasure'):
+                self._monitor_op = np.greater
+                self._best = -np.Inf
+            else:
+                self._monitor_op = np.less
+                self._best = np.Inf
+
+    def on_epoch_end(self, tuner: 'BaseTuner'):
+        """
+        Called at the end of the training epoch.
+        """
+        if self._monitor == 'loss':
+            self._save_model(tuner)
+            self._train_losses = []
+
+    def on_train_batch_end(self, tuner: 'BaseTuner'):
+        self._train_losses.append(tuner.state.train_loss)
+
+    def on_val_end(self, tuner: 'BaseTuner'):
```

Comment:
I think we only need to save in `on_epoch_end`

Reply:
what if we monitor validation loss and I want the best model after that epoch's validation

Reply:
`on_epoch_end` occurs after `on_val_end`, so there you will have both training and validation loss available for comparison